spring-integration-stream/src/main/java/org/springframework/integration/stream/CharacterStreamReadingMessageSource.java

```diff
@@ -57,20 +102,31 @@ else if (bufferSize > 0) {
 		else {
 			this.reader = new BufferedReader(reader);
 		}
+		this.blockToDetectEOF = blockToDetectEOF;
 	}
 
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
+		this.applicationEventPublisher = applicationEventPublisher;
+	}
 
+	@Override
 	public String getComponentType() {
 		return "stream:stdin-channel-adapter(character)";
 	}
 
+	@Override
 	public Message<String> receive() {
 		try {
 			synchronized (this.monitor) {
-				if (!this.reader.ready()) {
+				if (!this.blockToDetectEOF && !this.reader.ready()) {
 					return null;
 				}
 				String line = this.reader.readLine();
+				if (line == null && this.applicationEventPublisher != null && !this.eventPublished) {
+					this.applicationEventPublisher.publishEvent(new StreamClosedEvent(this));
+					this.eventPublished = true;
```

Comment:
I'm not sure in this `eventPublished` state.
As for me that looks natural to receive `StreamClosedEvent` whenever I call this `receive()`. 
Any call to closed stream should be ended with the same result.
In our case it is an `ApplicationEvent`.
WDYT?